core/primitives/src/shard_layout.rs

```diff
@@ -0,0 +1,143 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use byteorder::{LittleEndian, ReadBytesExt};
+use near_primitives_core::hash::hash;
+use near_primitives_core::types::ShardId;
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering::Greater;
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub enum ShardLayout {
+    V0(ShardLayoutV0),
+    V1(ShardLayoutV1),
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV0 {
+    /// Map accounts evenly across all shards
+    num_shards: NumShards,
+    /// Parent shards for the shards, useful for constructing states for the shards.
+    /// None for the genesis shard layout
+    parent_shards: Option<Vec<ShardId>>,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV1 {
+    /// num_shards = fixed_shards.len() + boundary_accounts.len() + 1
+    /// Each account and all subaccounts map to the shard of position in this array.
+    fixed_shards: Vec<AccountId>,
+    /// The rest are divided by boundary_accounts to ranges, each range is mapped to a shard
+    boundary_accounts: Vec<AccountId>,
+    /// Parent shards for the shards, useful for constructing states for the shards.
+    /// None for the genesis shard layout
+    parent_shards: Option<Vec<ShardId>>,
+}
+
+impl ShardLayout {
+    pub fn default(num_shards: NumShards) -> Self {
+        Self::V0(ShardLayoutV0 { num_shards, parent_shards: None })
+    }
+
+    pub fn v1(
+        fixed_shards: Vec<AccountId>,
+        boundary_accounts: Vec<AccountId>,
+        parent_shards: Option<Vec<ShardId>>,
+    ) -> Self {
+        Self::V1(ShardLayoutV1 { fixed_shards, boundary_accounts, parent_shards })
+    }
+    #[inline]
+    pub fn parent_shards(&self) -> &Option<Vec<ShardId>> {
+        match self {
+            Self::V0(v0) => &v0.parent_shards,
+            Self::V1(v1) => &v1.parent_shards,
+        }
+    }
+
+    #[inline]
+    pub fn num_shards(&self) -> NumShards {
+        match self {
+            Self::V0(v0) => v0.num_shards,
+            Self::V1(v1) => (v1.fixed_shards.len() + v1.boundary_accounts.len() + 1) as NumShards,
+        }
+    }
+}
+
+pub fn account_id_to_shard_id(account_id: &AccountId, shard_layout: &ShardLayout) -> ShardId {
+    match shard_layout {
+        ShardLayout::V0(ShardLayoutV0 { num_shards, .. }) => {
+            let mut cursor = Cursor::new(hash(&account_id.clone().into_bytes()).0);
+            cursor.read_u64::<LittleEndian>().expect("Must not happened") % (num_shards)
+        }
+        ShardLayout::V1(ShardLayoutV1 { fixed_shards, boundary_accounts, .. }) => {
+            for (shard_id, fixed_account) in fixed_shards.iter().enumerate() {
+                if is_top_level_account(fixed_account, account_id) {
+                    return shard_id as ShardId;
+                }
```

Comment:
I am a bit confused by this. Could you explain the motivation here?

Reply:
The comments are in ShardLayoutV1.

For this shard layout, there are some fixed shards, each fixed shard is mapped to an account and all subaccounts of this account. The rest of accounts are divided by boundary_accounts. For example, For `ShardLayoutV1{
fixed_shards: ["aurora"],
boundary_accounts: ["near"],
}`
Account "aurora" and all its subaccounts will be mapped to shard_id 0. For the rest of accounts, accounts <= "near" will be mapped to shard_id 1 and accounts > "near" will be mapped shard_id 2.

Reply:
> Account "aurora" and all its subaccounts will be mapped to shard_id 0

Is this actually what we want? Also adding a comment here will be helpful :)

Reply:
Sorry I missed this comment. 

I thought this is what we want, but maybe I misunderstood something. Is you question here regarding whether we should map all subaccounts of "aurora" to shard_id 0? I'll add a comment here.

Reply:
> Is you question here regarding whether we should map all subaccounts of "aurora" to shard_id 0?

Yes. I don't think we've agreed on that. Might require more discussion. @djsatok is there plan to create aurora subaccounts? If so, is it important that they are on the same shard with `aurora`?